### PR TITLE
ATO-1099: Update session when max age has expired

### PR DESCRIPTION
**Pull request by billrdunn**
## What
There is a specific set of updates required to the Orch session in relation to max age when all the following conditions are true:
- There is an existing Orch session
- Max age is enabled for the client
- Max age has expired

The updates to the Orch session are:
- Update the session ID of the existing (previous) Orch session. The session ID is updated as a security measure. Note that the previous session is not deleted (contrary to the non max age session update journey) because it is required after Authentication. The value of internalCommonSubjectIdentifier from the previous session will be compared against the value in the userinfo response, to confirm that the reauthenticated user is the same as the original user. Also the TTL of the previous session is bumped to ensure it is accessible after Authentication.
- Update the existing Orch session with authenticated=false (to force reauthentication) and add the new session ID for the previous session (to query ICSI from the previous session as described above).

Implementation has been feature flagged.

## Testing
Successfully run through the following journeys in dev, with both query parameter and request object requests:
- max age not set, sign in 2FA journey, claims returned
- max age not set, no sign in journey, claims returned
- max age 0 forces reauthentication
- max age 1234 does not force reauthentication
- max age 5 forces reauthentication
- max age -3 fails with appropriate error message
- max age “abc” fails with appropriate error message

### CODE REVIEW

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (Ryan-Andrews99) [inline oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:726]: Do we want this to be a warn instead of an error? Assuming this throwing is RP provided a junk value, in which case we could end up with noisy error alarms

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsAuthorizeValidator.java
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsAuthorizeValidator.java:144]: Actually nimbus throws in `AuthenticationRequest.parse(requestParameters)` before we do this validation. But yes we can catch the exception there and return something more useful to the RP.

Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsAuthorizeValidator.java:144]: Turns out we already return INVALID_REQUEST when the request params can't be parsed, so nothing to do here I don't think

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:642]: Works for me! Have updated

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:731]: hmm not sure what you mean

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:726]: RP providing a junk value should be caught earlier on. I can't imagine a case when this exception would be thrown but thought it would be good to add it just in case. Error logging so we're aware if max age isn't functioning as intended

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java]: Updated :)

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (Ryan-Andrews99) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:731]: Hmm yeah bit of an unclear comment, sorry! 😅 I'm not really a fan of the nimbus method `getMaxAge` returning -1 when not present  and we're kind of leaning into it here. Was thinking if we used an optional, the output is a bit clearer/ doesn't require context that nimbus does something weird
```
 private Optional<Long> getMaxAge(AuthenticationRequest authRequest) {
        Optional<Long>  maxAge = Optional.empty();
        try {
            var maxAgeParam = authRequest.getMaxAge();
            //Nimbus returns -1 if maxAge parameter is not present
            if (maxAgeParam != -1) {
                maxAge = Optional.of((long) maxAgeParam);
            }
        } catch (Exception e) {
            LOG.error(
                    "Failed to parse MaxAge param form Auth request, assuming no max age parameter ");
        }
        return  maxAge;
    }
```

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (Ryan-Andrews99) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:726]: Ah makes sense

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:740]: Ah yeah i see what you mean, LGTM

Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java]: Updated, changed a couple of things in your example. Also updated `maxAgeExpired()` 

and `if (authTime > timeNow)` check is new too

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsAuthorizeValidator.java
Comment (CarlyG55) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsAuthorizeValidator.java:1]: We should, yes!

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (CarlyG55) [inline oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:732]: I wonder if we want to be pulling these methods out this file, as it's getting so long. I created an `AuthorisationHelper` file a little while back - might be nice to move them into there?

Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java]: As discussed, will add this to tech debt in order to get this deployed and unblock other max age work

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsAuthorizeValidator.java
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsAuthorizeValidator.java:1]: I thought this would Just Work but it doesn't, so looking into it

Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsAuthorizeValidator.java:1]: Works now :)

Comment (Ryan-Andrews99) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsAuthorizeValidator.java:1]: Nice!

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/Session.java
Comment (billrdunn) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/Session.java:56]: Not totally convinced on whether this is needed. Seems like it will be only used for one specific case. Could all these fields just be set in `copySessionForMaxAge` instead?

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (billrdunn) [inline oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:615]: Could this be in the `else`? As there's no point calculating it if the session is empty

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (billrdunn) [inline oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:751]: `updateSharedSessionDueToMaxAgeExpiry` to match the other method?

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (billrdunn) [inline oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:647]: Does this logging cover both the shared and orch sessions? In which case might be better to use `newSessionId` instead of `orchSession.getSessionId()` and update the messages to make that clear

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (billrdunn) [inline oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:659]: I'm OK with this but for readability feels a little odd to be using `sessionToBeUpdated` after it's been updated. Could replace `sessionToBeUpdated` with `session` on lines after 659

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (Ryan-Andrews99) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:615]: Ooh yeah you're right

Comment (Ryan-Andrews99) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:615]: Have moved this and the timeNow variables into the else block

**Review by isaac-GDS — DISMISSED** [review]

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/Session.java
Comment (Ryan-Andrews99) [reply orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/Session.java:56]: Hmm would you prefer a _plain_ copy constructor with the fields being modified in the method instead?

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (Ryan-Andrews99) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:648]: Think I've updated this to do what you've described

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java
Comment (Ryan-Andrews99) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthorisationHandler.java:660]: Updated this

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/Session.java
Comment (Ryan-Andrews99) [reply orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/Session.java:56]: Ah yeah I was um-ing and ah-ing about this bit - happy to change!

**Review by CarlyG55 — APPROVED** [review]

**Review by billrdunn — COMMENTED** [review]
I can't approve as I'm a contributor but your changes LGTM

**Review by isaac-GDS — APPROVED** [review]

**Pull request merged**
